### DISCUSSION TITLE
Why does this fragment of code not work? (`IConfiguration`, `IChangeToken`)

### DISCUSSION THREAD

**Discussion by DrkWzrd** [dotnet/runtime#90350] [Q&A] [answered]
I'm driving myself crazy because this must work, right?

This run is a test in a desktop environment.

```C#
[Fact]
public void FileChangeShouldUpdateConfiguration()
{
	var cBuilder = new ConfigurationBuilder();
	cBuilder.AddJsonFile("config.json", false, true);
	var config = cBuilder.Build();
	var token = config.GetReloadToken();
	token.RegisterChangeCallback((c) =>
	{
		//this fragment is never called, and the configuration never reloads on his own
	}, null);

	var value = config["database_config:port"];

	var jsonOriginalText = File.ReadAllText($"{Environment.CurrentDirectory}/config.json");
	
        var jobject = JsonNode.Parse(jsonOriginalText);
	jobject!["database_config:port"] = Random.Shared.Next(0, 10000);

	File.WriteAllText($"{Environment.CurrentDirectory}/config.json", jobject.ToJsonString());

	var jsonNewText = File.ReadAllText($"{Environment.CurrentDirectory}/config.json");

	var newVal = config["database_config:port"];

	Assert.NotEqual(value, newVal);

}
```

No exceptions, no problems, only the configuration never updates itself. If I call `Reload()` then it works.

**Comment by KeterSCP** [comment, marked as answer]
Seems like ConfigurationBuilder uses `FileSystemWatcher` to detect changes on configs. And it can fire events with some delay and not immediately after a file was written. You can see that by simply adding `Thread.Sleep(500);` after `File.WriteAllText`

**Reply by DrkWzrd** [reply]
Despite of that being annoying, is the only way for the test to work always. Using a Thread.Sleep or Task.Delay (in async) with a number bigger (100-200 more) than the `ReloadDelay`. Even when the `ReloadDelay` is 0, the race must be controlled.

In non-test code, the callback is called, eventually.

This should be more clear in the docs, or it should be more instantaneous.
